File: Stationery-web-client/src/types/product.ts
import { Review } from './comment'
import { ProductPromotion } from './promotion'

interface Size {
  sizeId: string
  name: string
  priority: number
}

export interface ProductDetail {
  productDetailId: string
  name: string
  slug: string
  originalPrice: number
  stockQuantity: number
  soldQuantity: number
  discountPrice: number
  size: Size
  color: Color
  description: string
  productPromotions: ProductPromotion[]
  images: Image[] | null
  productId: string
  createdAt: string
  hidden: boolean
}

interface Image {
  imageId: string
  url: string
  priority: number
}

interface Color {
  colorId: string
  name: string
  hex: string
}

interface ProductColor {
  productColorId: string
  color: Color
  productDetails: ProductDetail[]
  images: Image[]
  sizes: Size[]
}

interface FetchColor {
  colorId: string
  hex: string
  slug: string
}

interface Product {
  productId: string
  description: string
  slug: string
  name: string
  category: {
    categoryId: string
    categoryName: string
  }
  soldQuantity: number
  quantity: number
  totalRating: number
  productDetail: ProductDetail
  fetchColor: FetchColor[]
  img: string
  createdAt: string
  hidden: boolean
}

interface ProductDeatilResponse extends Product {
  image: Image[]
  reviews: Review[]
  hidden: boolean
}

interface ListProductDetail extends Omit<Product, 'productDetail'> {
  categoryId: string
  productDetails: ProductDetail[]
}

interface ProductDetailResponse {
  code: number
  result: ProductDetail[]
}

interface ProductDetailForm {
  productDetailId?: string
  slug: string
  originalPrice: number
  stockQuantity: number
  discountPrice: number
  name: string
  sizeId: string
  colorId: string
  images?: {
    file: File
    url: string
    imageId?: string
  }[]
  deleteImages?: string[]
}
interface ProductForm {
  name: string
  description: string
  slug: string
  categoryId: string
  productDetails: ProductDetailForm[]
}
export type {
  ProductDetailForm,
  ProductForm,
  Product,
  Color,
  ProductColor,
  ProductDeatilResponse,
  Image,
  Size,
  FetchColor,
  ListProductDetail,
  ProductDetailResponse
}

File: stationery-be-server/src/main/java/com/project/stationery_be_server/entity/Product.java
package com.project.stationery_be_server.entity;

import com.fasterxml.jackson.annotation.*;
import com.project.stationery_be_server.dto.response.ColorSlugResponse;
import com.project.stationery_be_server.listener.ProductEntityListener;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(ProductEntityListener.class)
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "product_id")
    private String productId;

    @Column(name = "description", length = 500)
    private String description;

    @Column(name = "slug", length = 100)
    private String slug;
    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    @JsonIgnoreProperties({"icon", "bgColor", "products"})
    private Category category;

    private String name;

    @OneToMany(mappedBy = "product",fetch = FetchType.LAZY)
    @JsonManagedReference
    private List<Review> reviews;

    @Column(name = "total_rating")
    private Double totalRating;

    @Column(name="sold_quantity")
    private Integer soldQuantity;

    @Column(name="quantity")
    private Integer quantity;
    @Column(name = "created_at", nullable = false, updatable = false)
    @CreationTimestamp
    private LocalDateTime createdAt;

    @OneToMany(mappedBy = "product", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<ProductDetail> productDetails;

    @OneToMany(mappedBy = "product",fetch = FetchType.LAZY)
    @JsonIgnore
    List<Image> images;

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "default_pd", referencedColumnName = "product_detail_id")
    private ProductDetail productDetail;

    @Column(name="hidden",columnDefinition = "BOOLEAN DEFAULT FALSE")
    private boolean hidden =false;
    @Transient
    private List<ColorSlugResponse> fetchColor;
    @Transient
    private String img;
}
